sweep/sweeper.go

```diff
@@ -248,91 +271,57 @@ func (s *UtxoSweeper) collector() {
 			if pending {
 				log.Debugf("Already pending input %v received",
 					outpoint)
-
 				// Add additional result channel to signal spend
 				// of this input.
 				pendInput.listeners = append(
 					pendInput.listeners, input.resultChan)
-
 				continue
 			}
-
 			// Create a new pendingInput and initialize the
 			// listeners slice with the passed in result channel. If
 			// this input is offered for sweep again, the result
 			// channel will be appended to this slice.
 			pendInput = &pendingInput{
-				listeners: []chan Result{input.resultChan},
-				input:     input.input,
-				state:     stateNew,
+				listeners:        []chan Result{input.resultChan},
+				input:            input.input,
+				minPublishHeight: bestHeight,
 			}
 			s.pendingInputs[outpoint] = pendInput
 
-			size, err := getInputWitnessSizeUpperBound(input.input)
-			if err != nil {
-				err = fmt.Errorf("cannot determine size: %v",
-					err)
-
-				s.signalAndRemove(&outpoint, Result{Err: err})
-
-				continue
-			}
-			pendInput.witnessSizeUpperBound = size
-
 			// Start watching for spend of this input, either by us
 			// or the remote party. The possibility of a remote
 			// spend is the reason why just registering for conf. of
 			// our sweep tx isn't enough.
-			cancel, err := s.registerSpendNtfn(input.input)
+			cancel, err := s.registerSpendNtfn(
+				outpoint,
+				input.input.SignDesc().Output.PkScript,
+				input.input.HeightHint(),
+			)
 			if err != nil {
-				err = fmt.Errorf("cannot watch for spend: %v",
+				err = fmt.Errorf(
+					"cannot watch for spend: %v",
 					err)
 
-				s.signalAndRemove(&outpoint, Result{Err: err})
+				s.signalAndRemove(&outpoint,
+					Result{Err: err})
 
 				continue
 			}
 			pendInput.ntfnRegCancel = cancel
 
-			// Lookup if there is already a confirmed or unconfirmed
-			// tx of ours that spends this input.
-			spendingTx, err := s.cfg.HasSpendingTx(outpoint)
-			if err != nil {
-				log.Errorf("HasSpendingTx %v: %v", outpoint,
-					err)
-
-				// TODO: Re-enable code below once HasSpendingTx
-				// is reliable. For now, just continue as if
-				// there is no spending tx.
+			// Check to see if with this new input a sweep tx can be
+			// formed.
+			s.scheduleSweep(bestHeight)
 
-				// s.signalAndRemove(&outpoint, Result{Err: err})
-				// continue
-			}
-
-			if spendingTx != nil {
-				spendingTxHash := spendingTx.TxHash()
-				log.Debugf("Already swept input %v received "+
-					"(tx: %v)", outpoint, spendingTxHash)
-
-				// Save spending tx to find sibling inputs in
-				// case of remote spend.
-				pendInput.txHash = spendingTxHash
-				pendInput.state = statePublished
-
-				continue
+		case spend := <-s.spendChan:
```

Comment:
For actual testing, seems like this can be achieved with a mocked out `ChainNotifier`?

Reply:
I thought about it, but couldn't find a way to wait for this event to happen.